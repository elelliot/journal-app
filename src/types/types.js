//Tipos de Acciones

export const types = {

    login:'[Auth] Login',
    logout:'[Auth] Logout',

    //Se puede poner los nombres que se quiera pero que quede muy claro cual es el reducer que ejecuta la accion
    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',

    //con esto cambiamos el estado de loading a false o true seg√∫n necesitamos
    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading',

    //Para las notas
    notesAddNew: '[Notes] New Note',
    notesActive: '[Notes] Set active note',
    notesLoad: '[Notes] Load Notes',
    notesUpdate: '[Notes] Updated note',
    notesFileUrl: '[Notes] Updated image url',
    notesDelete: '[Notes] Delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning',

}