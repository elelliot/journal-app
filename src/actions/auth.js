

import { getAuth, signInWithPopup, createUserWithEmailAndPassword, signInWithEmailAndPassword, updateProfile, signOut } from "firebase/auth";
import { Toast } from "../alerts/alert";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { noteLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

//Alerts ver en alert.js
const welcome = () => { Toast.fire({
    iconColor: '#41D70C',
    icon: 'success',
    title: 'Welcome, what do we have for today? ^u^/'
  })};

const loginErr = () => {Toast.fire({
    icon: 'error',
    title: 'User or Password might be wrong or don\'t exist, try again ^u^'
  })};


const emailUsed = () => {Toast.fire({
    icon: 'error',
    title: 'Email already used'
  })};
//actions


//acción que dispara otra acción(login) cuando el setTimeout(tarea asíncrona) se resuelve
export const startLoginEmailPassword = (email, password) => {

    //al usar el dispatch de redux, la funcion retorna un callback ()=> ,y uno de sus args es el dispatch de Thunk  (todo adentro del callback se ejecuta)
    //se espera que se use el dispatch para trabajar los datos que llegan de una tarea async

    //se pueden realizar fetch, posteo de archivos o incluso dispatch de otra acción dentro del callback
    return (dispatch) => {
    
        //loading: true
        dispatch( startLoading() );

        //Para loguearse con email y password
        const auth = getAuth();
           return signInWithEmailAndPassword(auth, email, password)
            .then( ({user}) => {
                //console.log(user)
                //Mandamos uid y displayName al state cuando se hace login
                //dispatch( login( user.uid, user.displayName) );
                
                //PA LAS PRUEBAS puse el return y el dispatch del login(con esto se dispararía 2 veces en la app):
                dispatch( login( user.uid, user.displayName) );
                
                welcome()
            })
            .catch( ({ message }) => {
                
                //Args: Title, HTML, Icon... Alerta de error de Sweet Alert2
                //Swal.fire('Error', message, 'error');
                
                loginErr()
            })
            .finally( () => {
                //Para no tener que usar el finishLoading 2 veces podemos usarlo una vez aquí y en cualquiera de los 2 casos se llama
                //loading: false
                dispatch( finishLoading() );
            });
    }
}



//Registrar email
export const startRegisterWithEmail = ( email, password, name ) => {
    //en cuanto tenga el user grabado en firebase realizaré el dispatch
    return( dispatch ) => {

        const auth = getAuth();
        //en cuanto se crea el user esto intenta autenticar inmediatamente por eso da error por ahora.
        createUserWithEmailAndPassword(auth, email, password)
            .then( async ({ user }) => {
                
                //"displayName" es null usando esta función, solo se obtiene cuando se usa una red social como login
                //podemos asignar el name(el que damos en el form) al displayName con updateProfile
                await updateProfile(user,{displayName: name})
                
                //ya que el useEffect se ejecuta antes del updateProfile, el name es null, así que por ahora haremos el mismo dispatch con el nombre obtenido del update y tendremos 2 login al terminar el registro
                dispatch( login(user.uid, user.displayName ) )
                welcome()
            })
            .catch( ({ code } ) => {
                //desestructuramos message de error
                
                emailUsed()
            });
    }
}



//login con google 
export const startGoogleLogin = () =>{
    return (dispatch) =>{
        const auth = getAuth();
        //(aquí me tira el popup de login)
        signInWithPopup(auth, googleAuthProvider)
        //se obtiene el parametro por default y desestructuramos user del mismo
            .then( ({ user }) =>{
                //desestructuramos para obtener uid y displayName (user es solo una propiedad del objeto grande que te da la promise al cumplirse, por eso se ponen así en específico):
                //dispatch( login( user.uid, user.displayName ) );
                
                welcome()
            })
            .catch( ( { code } )  => {
                //por si se cierra el popup sin haber seleccionado usuario
                
                // const errorCode = error.code;
                // const errorMessage = error.message;
                //console.log(message);
                Toast.fire({
                    icon: 'error',
                    title: code
                  })
            });
    }
}


//El login que se usa en las otras 2 acciones
export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid, 
        displayName
    }
});


//Logout es Async por que hay que se usa una función firebase para esto y usa promesa

export const startLogout = () => {
    return async( dispatch ) => {
        const auth = getAuth();
        await signOut(auth)
            .then( ()=>{

                //limpiamos el user del state 
                dispatch( logout() );
                //limpiamos las notas del state
                dispatch( noteLogout() );
            })
            .catch( (e) => {
                console.log(e); //no suele pasar pero en caso de que se vaya el internet o algo
            })
    }
}

export const logout = () => ({
    type: types.logout,
});


//forma larga del return
// export const login = (uid, displayName) => {
//     return {
//         type: types.login,
//         payload: {
//             uid, 
//             displayName
//         }
//     }
// }


