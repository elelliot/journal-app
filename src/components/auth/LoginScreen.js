import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'

import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
    
    //igual que en el RegisterScreen, obtenemos el loading del ui del state de redux
    //esto va en el disabled = del botón de login
    const { loading } = useSelector( state => state.ui );
    
    //dispatch de acciones al store con redux 
    const dispatch = useDispatch();

    //Form
    const [ formValues, handleInputChange ] = useForm({
        email: 'elelliot@gmail.com',
        password: '123456'

    });

    const { email, password } = formValues;



    //Submit
    const handleLogin = (e) => {
        e.preventDefault();//para que no se refresque
        // console.log(email, password)
        //dispatch(login(123,'ElElliot'))

        //dispatch(el de Redux) de acción al store, recibe email y password en el actions/auth.js
        dispatch( startLoginEmailPassword(email, password) );

    }

    //botonazo de google
    const handleGoogleLogin = () =>{

        //como usamos signInwithPopup, no evitamos el refresh, pero para movil, sería apropiado manejarlo
        dispatch( startGoogleLogin() )
    }

    return (
        // Formulario V1
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className="auth__title">Login</h3>
            <form onSubmit ={ handleLogin }>
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="on"
                    value = { email }
                    onChange = { handleInputChange }
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="on"
                    value = { password }
                    onChange = { handleInputChange }
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={ loading }
                >
                    Login
                </button>

                {/* Botón de Google, se supone que el auth te dice que está en ese archivo, y luego iría la clase, pero something´s off*/}
                <div className="auth__social-networks">
                    <p>Login with social network</p>
                    <div 
                        className="google-btn"
                        onClick={ handleGoogleLogin }
                        >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link 
                    to="/auth/register"
                    className="link"
                >
                    Create new Account
                </Link>


            </form>
        </div>

    )
}
