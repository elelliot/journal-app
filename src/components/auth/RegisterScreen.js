import React from 'react';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';

import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmail } from '../../actions/auth';

export const RegisterScreen = () => {
    //dispatch de acciones al store con redux
    const dispatch = useDispatch();

    /*useSelector es un hook que dispara un callback en el cual tengo el state completo del store de Redux
    queremos agarrar el mensaje de error, como está en el ui lo extraemos con su nombre.*/
    const {msgError} = useSelector( state => state.ui );

    //Recordar que React redibuja y dispara el mensaje con cualquier cambio, y el msgError por defecto es null
    //console.log(msgError);


    //Form
    const [ formValues, handleInputChange ] = useForm({
        name:'Elliot',
        email:'elelliot@gmail.com',
        password: 123456,
        password2: 123456
    });
    const { name, email, password, password2 } = formValues;
    
    //Submit
    const handleRegister = (e) =>{
        //Evitamos Refresh y propagación por el URL
        e.preventDefault();
        
        if ( isFormValid() ){
            
            //console.log('Formulario Correcto');
            //dispatch para registrar a firebase con correo (debe ser diferente al de google)

            dispatch( startRegisterWithEmail(email, password, name) );
        }

    }

    /*
    Validamos Formulario (en lugar de hacer RegEx una alternatva sería usar la librería de npm Validator)
    el <input type="email"> no se usó, why?
    -dispatch al store de cada mensaje de error
    */
    const isFormValid = () => {
        
        if( name.trim().length === 0){
            //Si no hay nada escrito en el nombre
            dispatch( setError('Name is Required') );
            
            return false;
        } else if( !validator.isEmail( email ) ){
            //si el email no es válido (usamos la librería de validator para eso)
            dispatch( setError('Email not Valid') );
            
            //console.log('Email not Valid'); 

            return false;
        } else if ( password !== password2 || password.length < 5 ) {
            //Si el Password no coincide o si es menor a 5 car. 
            //QUEDA PENDIENTE CAMBIAR EL COMPORTAMIENTO (VER MI PREGUNTA EN UDEMY CLASE 251)
            dispatch( setError('Password should be at least 6 characters and Match each other') );
            
            //console.log('Password should be at least 6 characters and Match each other');
            return false;
        }
        //si es válido
        dispatch( removeError() ); 
        return true;
    }


    return (
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className="auth__title">Register</h3>

            
            {
                //Mensaje de Error del Formulario se maneja en la acción setError para usar Sweet Alert, 
                //ya que el msgError es null cuando queremos usarlo en el isFormValid y usarlo aquí rompe la app
                //si msgError es !== null muestra el mensaje de error
                msgError &&
                (
                    <div className="auth__alert-error">
                        { msgError }
                    </div>
                )
            }

            <form onSubmit ={ handleRegister }>
            <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={ name }
                    onChange = { handleInputChange }
                />
                
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange = { handleInputChange }
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="on"
                    value={ password }
                    onChange = { handleInputChange }
                />

                <input
                    type="password"
                    placeholder="Confirm Password"
                    name="password2"
                    className="auth__input"
                    autoComplete="on"
                    value={ password2 }
                    onChange = { handleInputChange }
                />


                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                    // disabled={ true }
                >
                    Register
                </button>

                <Link 
                    to="/auth/login"
                    className="link "
                >
                    Already Registered?
                </Link>


            </form>
        </div>
    )
}
