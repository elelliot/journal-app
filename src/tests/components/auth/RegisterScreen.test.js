import '@testing-library/jest-dom';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from 'react-router-dom'

import { RegisterScreen } from "../../../components/auth/RegisterScreen"
import { types } from '../../../types/types';



const middlewares = [thunk];
const mockStore = configureStore(middlewares);

//Le asignamos el estado que tendría el RegisterScreen, o al menos lo escencial
const initState = {
    auth: {},
    ui:{
        loading: false,
        msgError: null
    }
};
let store = mockStore( initState );
//store.dispatch = jest.fn();

const wrapper = mount(
        <Provider store = { store }>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
);

describe('Pruebas en <RegisterScreen />', () => {


    test('Debe mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();
    });


    test('Debe hacer el dispatch de la acción respectiva', () => {

        const emailField = wrapper.find('input[name="email"]');//seleccionamos un elemento por el name
        //console.log(emailField.exists())
        
        //Modificamos el campo, ahora el email está vacío en el registerForm
        emailField.simulate('change',{
            target:{
                value:'',
                name:'email'
            }
        });

        //Simulamos el submit del form con el campo de email vacío(igual se puede hacer con el prop que sería onSubmit)
        wrapper.find('form').simulate('submit',{
            preventDefault(){}
        });

        //Como el aviso de error es sincrono, no necesitamos el store.dispatch = jest.fn(), osea simular el async dispatch
        const actions = store.getActions();
        
        expect( actions[0] ).toEqual({
            type: types.uiSetError,
            payload: 'Email not Valid'
        });
    })


    test('Debe mostrar la alerta con el error (el div, no el Swal)', () => {

        const initState = {
            auth: {},
            ui:{
                loading: false,
                msgError: 'Email not valid (Test)'
            }
        };
        const store = mockStore( initState );
        
        const wrapper = mount(
                <Provider store = { store }>
                    <MemoryRouter>
                        <RegisterScreen />
                    </MemoryRouter>
                </Provider>
        );


        expect( wrapper.find('.auth__alert-error').exists() ).toBe( true );
        expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initState.ui.msgError );
    })
})