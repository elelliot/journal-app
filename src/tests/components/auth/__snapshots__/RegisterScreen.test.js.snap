// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen /> Debe mostrarse correctamente 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Elliot"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="elelliot@gmail.com"
    />
    <input
      autoComplete="on"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value={123456}
    />
    <input
      autoComplete="on"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value={123456}
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link "
      href="/auth/login"
      onClick={[Function]}
    >
      Already Registered?
    </a>
  </form>
</div>
`;
